BetterPasswordPlugin.inc.php

```diff
@@ -208,17 +199,109 @@ function checkPassword($hookName, $args) {
+	 */
+        
+        function blacklistSetting() {
+            $prevBlacklist = $this->getSetting(CONTEXT_SITE, 'betterPasswordBlacklistFiles');
+            $updateSettings = false;
+            $newBlacklist = array();
+            foreach ($this->getBlacklists() as $filename) {
+                    $newBlacklist[$filename] = sha1_file($filename);
+            }
+            if (is_null($prevBlacklist)) {
+                $this->updateSetting(CONTEXT_SITE, 'betterPasswordBlacklistFiles', $newBlacklist);
+                $updateSettings = true;
+            }
+            else {
+                if ($prevBlacklist != $newBlacklist) {
+                    $updateTempFile = $this->handleTempFile(false);
+                    if($updateTempFile) {
+                        $this->updateSetting(CONTEXT_SITE, 'betterPasswordBlacklistFiles', $blacklistFileSettingHash);
+                        $updateSettings = true;
+                    }
+                    else {
+                        return $updateSettings;
+                    }
+                }
+            }
+            return $updateSettings;
+        }
+        
+        /**
+         * Creates a temporary file to aggregate all the passwords from the blacklist files if the temporary file doesn't exist
+         * @return array The file path for the temporary file
+         */
+        function getTempFile() {
+            import('lib.pkp.classes.file.PrivateFileManager');
+            $fileMgr = new PrivateFileManager();
+            $tempFileDir = realpath($fileMgr->getBasePath()) . DIRECTORY_SEPARATOR . 'betterPassword';
+            if (!$fileMgr->fileExists($tempFileDir, 'dir')) {
+			$success = $fileMgr->mkdirtree($tempFileDir);
+			if (!$success) {
+                            // Files directory wrong configuration?
+                            return null;
+			}
+		}
+            if(!file_exists($tempFileDir . DIRECTORY_SEPARATOR . 'tempPassFile')) {
+                touch($tempFileDir . DIRECTORY_SEPARATOR . 'tempPassFile');
+                return $tempFileDir . DIRECTORY_SEPARATOR . 'tempPassFile';
+            }
+            else {
+                return $tempFileDir . DIRECTORY_SEPARATOR . 'tempPassFile';;
+            }
+        }
+        
+	/**
+	 * Handles the temporary file with the data of password blacklists.
+         * @param boolean $trustExistingFile Default true, only false when updating temp file
+	 * @return string|boolean|null temporary filename string, or boolean or null when there is an error, when $check is false return associative array
+	 */
+         function handleTempFile($trustExistingFile = true) {
+            $siteDao = DAORegistry::getDAO('SiteDAO');
+            $site = $siteDao->getSite();
+            $minLengthPass = $site->getMinPasswordLength();
+            $tempFilePath = $this->getTempFile();
+            if($trustExistingFile) {
+                $fpTemp = fopen($tempFilePath, 'a');
```

Comment:
Why are we writing to the file if we trust the existing contents?  Won't multiple calls to `handleTempFile(true)` now repeatedly append the contents of the blacklists onto one another?

Reply:
Something is not quite right about the logic if we $trustExistingFile, but it is empty and that causes us to want to append to it.  Should this method be creating the temp file as needed?

Reply:
When we call the getTempFile function we get either the path of an empty file or the path of the temp file if it exists already. The getTempFile function creates a temp file if there isn't one already.

Reply:
For the logic, we have 3 cases:

1. TrustExistingFile and temp file exists but is empty:
   In this case we want to append all the passwords to the temp file
2. We trust the existing file and temp file exists not empty:
    In this case we do not want to rewrite or append existing blacklist files to the temp file
3. We do not trust the existing file:
   Rewrite the temp file to contain the new blacklists

Temp File exists but empty

Reply:
If we don't trust the existing file, we need to overwrite it.  If we trust the existing file, we shouldn't modify it.  In what case should we trust the file, but still modify it?

Reply:
The first case where we trust the temp file but we need to modify it only occurs the first time we create the temp file. After that its only 2 cases, i.e, trust the temp file from that point forward or not.

Reply:
Case 1 is a cache miss which occurs before the blacklists have been processed at all.  How can this happen?  The processing of the blacklists occurs here:
https://github.com/ulsdevteam/pkp-betterPassword/pull/7/files#diff-ed7ee6c9e9c6cb7e4dda7918a5e9b147R164
and the cache check occurs here, only if we have a blacklist:
https://github.com/ulsdevteam/pkp-betterPassword/pull/7/files#diff-ed7ee6c9e9c6cb7e4dda7918a5e9b147R167